cmd/proxy.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/cloudsqlconn"
+	"github.com/spf13/cobra"
+)
+
+// proxyClient represents the state of the current instantiation of the proxy.
+type proxyClient struct {
+	cmd    *cobra.Command
+	dialer *cloudsqlconn.Dialer
+
+	// mnts is a list of all mounted sockets for this client
+	mnts []*socketMount
+	// serveCtx is used for serving data
+	serveCtx    context.Context
+	serveCancel context.CancelFunc
+}
+
+// newProxyClient completes the initial setup required to get the proxy to a "steady" state.
+func newProxyClient(ctx context.Context, cmd *cobra.Command, args []string) (*proxyClient, error) {
+	dialer, err := cloudsqlconn.NewDialer(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing dialer: %v", err)
+	}
+	pc := &proxyClient{cmd: cmd, dialer: dialer}
+
+	port := 5000 // TODO: figure out better port allocation strategy
+	for i, inst := range args {
+		m := newSocketMount(dialer, inst)
+		addr, err := m.Listen(ctx, "tcp4", net.JoinHostPort("", fmt.Sprint(port+i)))
```

Comment:
This will listen on all network interfaces and not just localhost, which is generally considered a security problem. Is there a TODO that should be here? Otherwise, I'd expect to see the host set to localhost.